Patch: 67_update_handle_old_versions.patch
Author: Bastian Kleineidam <calvin@debian.org>
Handle the case of upgrading from Mailman 2.0 where we have
pending subscriptions.
Index: bin/update
===================================================================
--- bin/update.orig	2006-08-15 15:14:50.000000000 +0800
+++ bin/update	2006-08-15 15:14:52.000000000 +0800
@@ -514,9 +514,11 @@
     file20 = os.path.join(mm_cfg.DATA_DIR, 'pending_subscriptions.db')
     file214 = os.path.join(mm_cfg.DATA_DIR, 'pending.pck')
     db = None
+    ver = None
     # Try to load the Mailman 2.0 file
     try:
         fp = open(file20)
+        ver = "20"
     except IOError, e:
         if e.errno <> errno.ENOENT: raise
     else:
@@ -528,6 +530,7 @@
         # Try to load the Mailman 2.1.x where x < 5, file
         try:
             fp = open(file214)
+            ver = "214"
         except IOError, e:
             if e.errno <> errno.ENOENT: raise
         else:
@@ -561,8 +564,12 @@
             # data[0] is the address being unsubscribed
             addrops_by_address.setdefault(data[0], []).append((key, val))
         elif op == Pending.SUBSCRIPTION:
-            # data[0] is a UserDesc object
-            addr = data[0].address
+            if ver == "20":
+                # data is tuple (emailaddr, password, digest)
+                addr = data[0]
+            else:
+                # data[0] is a UserDesc object
+                addr = data[0].address
             subs_by_address.setdefault(addr, []).append((key, val))
         elif op == Pending.RE_ENABLE:
             # data[0] is the mailing list's internal name
